R/subscript-tables.R

```diff
@@ -0,0 +1,118 @@
+#' @export
+`[.QTable` <- function(x, ..., drop = TRUE) {
+    # Use sys.call as match.call captures unmatched named arguments into ...
+    used.arguments <- names(sys.call())
+    input.name <- attr(x, "name")
+    if (!validArgumentNames(used.arguments, "drop"))
+        throwErrorOnlyNamed("drop", "[")
+    if ("drop" %in% used.arguments && !is.logical(drop))
+        stop("drop argument should be TRUE or FALSE")
+    called.args <- match.call(expand.dots = FALSE)
+    empty.ind <- providedArgumentEmpty(called.args, optional.arg = "drop")
+    # Catch empty input e.g. x[] or x[drop = TRUE/FALSE] (when ... is empty)
+    if (empty.ind) {
+        y <- NextMethod(x)
+        class(y) <- c("QTable", class(y))
+        return(y)
+    }
+    x.dim <- dim(x)
+    n.dim <- length(x.dim)
+    n.index.args <- nargs() - 1L - !missing(drop)
```

Comment:
`drop` will no longer ever be `missing`.

Reply:
Thanks, I'll update now

Reply:
Wait, it can still be missing, if the user doesn't provide it, it wont contribute to the count in `nargs()`. So this code needs to remain as is.

Reply:
```
> debug(`[.QTable`)
> x[1, 2, drop = TRUE]
debugging in: `[.QTable`(x, 1, 2, drop = TRUE)
debug at /home/jrw/flip/verbs/R/subscript-tables.R#2: {
    used.arguments <- names(sys.call())
    input.name <- attr(x, "name")
    if (!validArgumentNames(used.arguments, "drop"))
        throwErrorOnlyNamed("drop", "[")
    if ("drop" %in% used.arguments && !is.logical(drop))
        stop("drop argument should be TRUE or FALSE")
    called.args <- match.call(expand.dots = FALSE)
    empty.ind <- providedArgumentEmpty(called.args, optional.arg = "drop")
    if (empty.ind) {
        y <- NextMethod(x)
        class(y) <- c("QTable", class(y))
        return(y)
    }
    x.dim <- dim(x)
    n.dim <- length(x.dim)
    n.index.args <- nargs() - 1L - !missing(drop)
    if (n.index.args != 1 && n.dim != n.index.args)
        throwErrorTableIndexInvalid(input.name, n.dim, n.index.args)
    y <- NextMethod(x)
    class(y) <- c("QTable", class(y))
    attr(y, "name") <- input.name
    y
}
Browse[2]> nargs()
[1] 4
Browse[2]> missing(drop)
[1] FALSE
Browse[2]> Q
> x[1, 2]
debugging in: `[.QTable`(x, 1, 2)
debug at /home/jrw/flip/verbs/R/subscript-tables.R#2: {
    used.arguments <- names(sys.call())
    input.name <- attr(x, "name")
    if (!validArgumentNames(used.arguments, "drop"))
        throwErrorOnlyNamed("drop", "[")
    if ("drop" %in% used.arguments && !is.logical(drop))
        stop("drop argument should be TRUE or FALSE")
    called.args <- match.call(expand.dots = FALSE)
    empty.ind <- providedArgumentEmpty(called.args, optional.arg = "drop")
    if (empty.ind) {
        y <- NextMethod(x)
        class(y) <- c("QTable", class(y))
        return(y)
    }
    x.dim <- dim(x)
    n.dim <- length(x.dim)
    n.index.args <- nargs() - 1L - !missing(drop)
    if (n.index.args != 1 && n.dim != n.index.args)
        throwErrorTableIndexInvalid(input.name, n.dim, n.index.args)
    y <- NextMethod(x)
    class(y) <- c("QTable", class(y))
    attr(y, "name") <- input.name
    y
}
Browse[2]> missing(drop)
[1] TRUE
Browse[2]> nargs()
[1] 3
```

Reply:
My bad, good to know. thanks for catching!